File: GamesaveCloudCLI/Synchronizer.vb
Imports Google.Apis.Drive.v3
Imports System.Data
Imports System.Data.SQLite
Imports System.IO
Imports System.IO.Compression
Imports System.Reflection

Public Class Synchronizer

    Dim driveHelper As GoolgeDriveHelper

    Dim gamesaveRootFolder As Data.File
    Dim configFolder As Data.File
    Dim backupFolder As String
    Dim pathConfigFile As String
    Public version As String = "0.84"
    Dim logFile As StreamWriter

    ' configuration stored in sqlite
    Dim syncInterval As Integer
    Dim performBackup As Boolean

    Public Class Args
        Public command As String
        Public game_id As String
    End Class

    Public Sub Initialize()

        Dim pathAtual = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location)

        logFile = New StreamWriter(Path.Combine(pathAtual, "GamesaveCloud.log"), True)

        Dim myFile As New FileInfo(Path.Combine(pathAtual, "GamesaveCloud.log"))
        If myFile.Length > 0 Then
            logFile.WriteLine()
        End If

        Dim startTime = DateTime.Now
        Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + ": Connecting to Google Drive... ")

        driveHelper = New GoolgeDriveHelper(Assembly.GetExecutingAssembly().GetManifestResourceStream("GamesaveCloudCLI.drive_api.json"))

        gamesaveRootFolder = driveHelper.GetFolder("root", "GamesaveCloud")
        If gamesaveRootFolder Is Nothing Then
            gamesaveRootFolder = driveHelper.NewFolder("GamesaveCloud", "root")
        End If

        configFolder = driveHelper.GetFolder(gamesaveRootFolder.Id, "config")
        If configFolder Is Nothing Then
            configFolder = driveHelper.NewFolder("config", gamesaveRootFolder.Id)
        End If


        Dim pathConfigFolder = Path.Combine(pathAtual, "config")
        If Not System.IO.Directory.Exists(pathConfigFolder) Then
            Directory.CreateDirectory(pathConfigFolder)
        End If

        SyncPath(pathConfigFolder, configFolder.Id, False, "config")

        pathConfigFile = Path.Combine(pathConfigFolder, "GamesaveDB.db")

        If Not System.IO.File.Exists(pathConfigFile) Then
            Dim sReader As Stream = Assembly.GetExecutingAssembly().GetManifestResourceStream("GamesaveCloudCLI.GamesaveDB.db")
            Dim sWriter As Stream = New FileStream(pathConfigFile, FileMode.OpenOrCreate, FileAccess.Write, FileShare.None)
            BinaryStreamCopy(sReader, sWriter)
            sWriter.Flush()
            sWriter.Close()
            sReader.Close()

            SyncPath(pathConfigFolder, configFolder.Id, False, "config")
        End If

        backupFolder = Path.Combine(pathAtual, "backup")
        If Not System.IO.Directory.Exists(backupFolder) Then
            Directory.CreateDirectory(backupFolder)
        End If

        Dim endTime = DateTime.Now
        Dim secs As Double = endTime.Subtract(startTime).TotalSeconds

        Log($"connected in {secs.ToString("N2")} seconds" + Environment.NewLine)

        Dim sqlite_conn As SQLiteConnection
        Dim sqlite_cmd As SQLiteCommand
        Dim sqlite_datareader As SQLiteDataReader

        sqlite_conn = New SQLiteConnection("Data Source=" + pathConfigFile + ";Version=3;New=True;")
        sqlite_conn.Open()

        sqlite_cmd = sqlite_conn.CreateCommand()

        sqlite_cmd.CommandText = "select value from parameter where parameter_id = 'SYNC_INTERVAL'"
        sqlite_datareader = sqlite_cmd.ExecuteReader()
        If sqlite_datareader.Read() Then
            syncInterval = Convert.ToInt32(sqlite_datareader.GetString(0))
        Else
            syncInterval = 60000
        End If
        sqlite_datareader.Close()

        sqlite_cmd.CommandText = "select value from parameter where parameter_id = 'BACKUP'"
        sqlite_datareader = sqlite_cmd.ExecuteReader()
        If sqlite_datareader.Read() Then
            performBackup = Convert.ToInt32(sqlite_datareader.GetString(0)) = 1
        Else
            performBackup = True
        End If

        sqlite_conn.Close()

    End Sub

    Public Sub Sync(pgame_id As Integer)

        Dim sqlite_conn As SQLiteConnection
        Dim sqlite_cmd As SQLiteCommand
        Dim sqlite_datareader As SQLiteDataReader

        Dim startTime = DateTime.Now
        Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + $": *** Synchronizing... " + Environment.NewLine)

        sqlite_conn = New SQLiteConnection("Data Source=" + pathConfigFile + ";Version=3;New=True;")
        sqlite_conn.Open()

        'sqlite_cmd = sqlite_conn.CreateCommand()
        'sqlite_cmd.CommandText = "INSERT INTO people (first_name, last_name) VALUES ('Rui','Martins');"
        'sqlite_cmd.ExecuteNonQuery()

        sqlite_cmd = sqlite_conn.CreateCommand()

        If pgame_id = Nothing Then
            sqlite_cmd.CommandText = "select game_id, title from game where active = 1"
            sqlite_datareader = sqlite_cmd.ExecuteReader()
        Else
            sqlite_cmd.CommandText = "select game_id, title from game where active = 1 and game_id = @gameid"
            sqlite_cmd.Parameters.Add("@gameid", SqlDbType.Int).Value = pgame_id
            sqlite_datareader = sqlite_cmd.ExecuteReader()
        End If

        While (sqlite_datareader.Read()) ' Read() returns true if there is still a result line to read

            Dim game_id As Object
            Dim title As String
            Dim syncState As Integer = 0

            game_id = sqlite_datareader.GetValue(0)
            title = sqlite_datareader.GetString(1)

            Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + ": " + title + " ")

            Dim sqlite_cmd2 As SQLiteCommand
            Dim sqlite_datareader2 As SQLiteDataReader

            sqlite_cmd2 = sqlite_conn.CreateCommand()
            sqlite_cmd2.CommandText = "select savegame_id, path, machine from savegame where game_id = @gameid"
            sqlite_cmd2.Parameters.Add("@gameid", SqlDbType.Int).Value = game_id
            sqlite_datareader2 = sqlite_cmd2.ExecuteReader()

            While (sqlite_datareader2.Read())

                Dim savegame_id As Integer
                Dim path As String
                Dim machine As Integer

                savegame_id = sqlite_datareader2.GetValue(0)
                path = sqlite_datareader2.GetString(1)
                path = ReplaceEnvironmentVariables(path)
                machine = sqlite_datareader2.GetValue(2)

                Dim gameFolder As Data.File = driveHelper.GetFolder(gamesaveRootFolder.Id, game_id.ToString)
                Dim gamesaveFolder As Data.File = Nothing

                If gameFolder IsNot Nothing Then
                    gamesaveFolder = driveHelper.GetFolder(gameFolder.Id, savegame_id.ToString)
                    If gamesaveFolder IsNot Nothing Then
                        If Not Directory.Exists(path) Then
                            Directory.CreateDirectory(path)
                        End If
                    Else
                        If Directory.Exists(path) Then
                            gamesaveFolder = driveHelper.NewFolder(savegame_id.ToString, gameFolder.Id)
                        End If
                    End If
                Else
                    If Directory.Exists(path) Then
                        gameFolder = driveHelper.NewFolder(game_id.ToString, gamesaveRootFolder.Id)
                        gamesaveFolder = driveHelper.NewFolder(savegame_id.ToString, gameFolder.Id)
                    End If
                End If

                If gamesaveFolder IsNot Nothing Then

                    Dim syncResult = SyncPath(path, gamesaveFolder.Id, performBackup, game_id.ToString() + "_" + savegame_id.ToString() + "_" + DateTime.Now.ToString("ddMMyyyy_HHmmss"))
                    If syncResult = -1 Then
                        If (machine = 1) Then
                            ' update machine subfolders with the most recent when synchronizing from cloud
                            UpdateMachine(path)
                        End If
                        ' sync from cloud
                        syncState = 3
                    ElseIf syncResult = 1 Then
                        ' sync from local computer
                        syncState = 2
                    ElseIf syncResult = 0 Then
                        If syncState = 0 Then
                            ' already sync
                            syncState = 1
                        End If
                    End If

                End If

            End While
            sqlite_datareader2.Close()

            Select Case syncState
                Case 0
                    ' skipped: not in local and not in cloud and game active
                    Log("skipped" + Environment.NewLine)
                Case 1
                    Log("in sync" + Environment.NewLine)
                Case 2
                    Log("synched from local computer" + Environment.NewLine)
                Case 3
                    Log("synched from cloud" + Environment.NewLine)
            End Select

        End While
        sqlite_datareader.Close()
        sqlite_conn.Close()

        Dim endTime = DateTime.Now
        Dim secs As Double = endTime.Subtract(startTime).TotalSeconds
        Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + $": *** Synchronization completed in {secs.ToString("N2")} seconds" + Environment.NewLine)

    End Sub


    Public Sub Close()
        logFile.Close()
    End Sub


    Private Sub Log(ByVal text As String)
        Console.Write(text)
        logFile.Write(text)
    End Sub

    Private Sub BinaryStreamCopy(sReader As Stream, sWriter As Stream)

        ' Create a read buffer. The buffer length is less than or equal to the data length
        Dim dataLength = sReader.Length
        Dim bufferLength As Integer = CInt(If(dataLength >= 1024, 1024, dataLength))
        Dim buffer As Byte() = New Byte(bufferLength - 1) {}
        Dim read As Integer = 0

        While True
            read = sReader.Read(buffer, 0, buffer.Length)
            If read = 0 Then Exit While
            sWriter.Write(buffer, 0, read)
        End While

    End Sub

    ' Fully synchronizes a local and could folder
    ' Result: 0 no synch needed, 1 synchronizes from local, -1 synchronizes from cloud
    Private Function SyncPath(folderPath As String, folderId As String, backup As Boolean, backupName As String) As Integer

        Dim syncResult As Integer = 0
        Dim totalFilesDrive As Integer = 0
        Dim lastModifiedDrive As Date = driveHelper.LastModifiedDate(folderId, totalFilesDrive)
        Dim totalFilesLocal As Integer = 0
        Dim lastModifiedLocal As Date = driveHelper.LocalLastModifiedDate(folderPath, totalFilesLocal)

        If lastModifiedDrive <> Nothing Then
            If lastModifiedLocal <> Nothing Then
                If lastModifiedDrive.ToString("yyyyMMdd HHmmss") <> lastModifiedLocal.ToString("yyyyMMdd HHmmss") Or totalFilesDrive <> totalFilesLocal Then
                    If lastModifiedDrive > lastModifiedLocal Then
                        If backup Then
                            ZipFile.CreateFromDirectory(folderPath, Path.Combine(backupFolder, backupName) + ".zip", CompressionLevel.Optimal, True)
                        End If
                        driveHelper.SyncFromDrive(folderPath, folderId)
                        syncResult = -1
                    Else
                        driveHelper.SyncFromLocal(folderPath, folderId)
                        syncResult = 1
                    End If
                End If
            Else
                If backup Then
                    ZipFile.CreateFromDirectory(folderPath, Path.Combine(backupFolder, backupName) + ".zip", CompressionLevel.Optimal, True)
                End If
                driveHelper.SyncFromDrive(folderPath, folderId)
                syncResult = -1
            End If
        Else
            If lastModifiedLocal <> Nothing Then
                driveHelper.SyncFromLocal(folderPath, folderId)
                syncResult = 1
            End If
        End If

        Return syncResult

    End Function

    Private Function ReplaceEnvironmentVariables(sPath As String) As String

        Dim pos1 = sPath.IndexOf("%")
        If pos1 = -1 Then
            Return sPath
        End If
        Dim pos2 = sPath.IndexOf("%", pos1 + 1)
        If pos2 = -1 Then
            Return sPath
        End If

        Dim variable = sPath.Substring(pos1, pos2 - pos1 + 1)
        Return sPath.Replace(variable, Environment.GetEnvironmentVariable(variable.Replace("%", "")))

    End Function

    Private Sub UpdateMachine(folderPath As String)

        Dim lastModified As Date = Nothing
        Dim lastModifiedPath As String = Nothing
        Dim folderEntries As String() = Directory.GetDirectories(folderPath)
        For Each folderEntry In folderEntries
            Dim totalFiles As Integer = 0
            Dim folderLastModified As Date = driveHelper.LocalLastModifiedDate(folderEntry, totalFiles)
            If lastModified = Nothing Or folderLastModified > lastModified Then
                lastModified = folderLastModified
                lastModifiedPath = folderEntry
            End If
        Next

        For Each folderEntry In folderEntries
            If folderEntry <> lastModifiedPath Then
                Directory.Delete(folderEntry, True)
                Directory.CreateDirectory(folderEntry)
                FileIO.FileSystem.CopyDirectory(lastModifiedPath, folderEntry, True)
            End If
        Next

    End Sub


End Class

File: GamesaveCloudCLI/Synchronizer.cs
﻿using GamesaveCloudLib;
using Microsoft.Graph.Models;
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SQLite;
using System.IO;
using System.IO.Compression;
using System.Linq;
using System.Reflection;
using System.Runtime.Versioning;

namespace GamesaveCloudCLI
{

    public class Synchronizer
    {

        public static readonly List<string> CloudServices = new() { "googledrive", "onedrive" };
        public static readonly List<string> SyncDirections = new() { "auto", "tocloud", "fromcloud" };

        private ICloudDriveHelper driveHelper;
        private ICloudFile gamesaveRootFolder;
        private ICloudFile configFolder;
        private string backupFolder;
        private string pathConfigFile;
        public string version = "0.9";
        private StreamWriter logFile;
        public string cloudService;

        private DateTime sessionStartTime;

        // configuration stored in sqlite
        private bool performBackup;

        public class Args
        {
            public string command;
            public string game_id;
        }

        [SupportedOSPlatform("windows")]
        public void Initialize(string cloudService)
        {
            sessionStartTime = DateTime.Now;

            this.cloudService = cloudService;

            string pathAtual = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);

            string pathLog = Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location), "logs");

            Directory.CreateDirectory(pathLog);

            logFile = new StreamWriter(Path.Combine(pathLog, "GamesaveCloud.Log"), true);

            //var myFile = new FileInfo(Path.Combine(pathLog, "GamesaveCloud.Log"));
            //if (myFile.Length > 0L)
            //{
            //    logFile.WriteLine();
            //}

            var startTime = DateTime.Now;
            Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + ": Connecting to " + cloudService + "... " + Environment.NewLine);

            driveHelper = cloudService switch
            {
                "googledrive" => new GoolgeDriveHelper(),
                "onedrive" => new OneDriveHelper(),
                _ => new OneDriveHelper(),
            };
            gamesaveRootFolder = driveHelper.GetFolder("root", "GamesaveCloud");
            gamesaveRootFolder ??= driveHelper.NewFolder("GamesaveCloud", "root");

            var endTime = DateTime.Now;
            double secs = endTime.Subtract(startTime).TotalSeconds;

            Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + ": Connected with " + driveHelper._username + $" in {secs:N2} seconds" + Environment.NewLine);

            configFolder = driveHelper.GetFolder(gamesaveRootFolder.Id, "config");
            configFolder ??= driveHelper.NewFolder("config", gamesaveRootFolder.Id);

            string pathConfigFolder = Path.Combine(pathAtual, "config");
            if (!Directory.Exists(pathConfigFolder))
            {
                Directory.CreateDirectory(pathConfigFolder);
            }

            var dbSync = SyncPath(pathConfigFolder, configFolder.Id, false, "config", false, null, "auto");

            pathConfigFile = Path.Combine(pathConfigFolder, "GamesaveDB.db");

            if (!File.Exists(pathConfigFile))
            {
                string resourceName = Assembly.GetExecutingAssembly().GetManifestResourceNames().Single(str => str.EndsWith("GamesaveDB.db"));
                var sReader = Assembly.GetExecutingAssembly().GetManifestResourceStream(resourceName);
                Stream sWriter = new FileStream(pathConfigFile, FileMode.OpenOrCreate, FileAccess.Write, FileShare.None);
                BinaryStreamCopy(sReader, sWriter);
                sWriter.Flush();
                sWriter.Close();
                sReader.Close();

                SyncPath(pathConfigFolder, configFolder.Id, false, "config", false, null, "auto");

                Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + ": Database initialized and synched to " + cloudService + Environment.NewLine);
            }
            else
            {
                switch (dbSync)
                {
                    case 1:
                        Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + ": Database synched from local computer" + Environment.NewLine);
                        break;
                    case -1:
                        Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + ": Database synched from " + cloudService + Environment.NewLine);
                        break;
                    default:
                        Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + ": Database in sync" + Environment.NewLine);
                        break;
                }

            }

            backupFolder = Path.Combine(pathAtual, "backup");
            if (!Directory.Exists(backupFolder))
            {
                Directory.CreateDirectory(backupFolder);
            }

            SQLiteConnection sqlite_conn;
            SQLiteCommand sqlite_cmd;
            SQLiteDataReader sqlite_datareader;

            sqlite_conn = new SQLiteConnection("Data Source=" + pathConfigFile + ";Version=3;New=True;");
            sqlite_conn.Open();

            sqlite_cmd = sqlite_conn.CreateCommand();

            sqlite_cmd.CommandText = "select value from parameter where parameter_id = 'BACKUP'";
            sqlite_datareader = sqlite_cmd.ExecuteReader();
            if (sqlite_datareader.Read())
            {
                performBackup = Convert.ToInt32(sqlite_datareader.GetString(0)) == 1;
            }
            else
            {
                performBackup = true;
            }
            sqlite_datareader.Close();
            sqlite_conn.Close();

        }

        public void Sync(int? gameId, string gameTitle, string syncDirection)
        {
            SQLiteConnection sqlite_conn;
            SQLiteCommand sqlite_cmd;
            SQLiteDataReader sqlite_datareader;

            var startTime = DateTime.Now;
            Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + $": *** Synchronizing... " + Environment.NewLine);

            sqlite_conn = new SQLiteConnection("Data Source=" + pathConfigFile + ";Version=3;New=True;");
            sqlite_conn.Open();

            // sqlite_cmd = sqlite_conn.CreateCommand()
            // sqlite_cmd.CommandText = "INSERT INTO people (first_name, last_name) VALUES ('Rui','Martins');"
            // sqlite_cmd.ExecuteNonQuery()

            sqlite_cmd = sqlite_conn.CreateCommand();

            if (gameId == 0)
            {
                sqlite_cmd.CommandText = "select game_id, title from game where active = 1";
                sqlite_datareader = sqlite_cmd.ExecuteReader();
            }
            else if (gameId != null)
            {
                sqlite_cmd.CommandText = "select game_id, title from game where active = 1 and game_id = @gameid";
                sqlite_cmd.Parameters.AddWithValue("@gameid", gameId);
                sqlite_datareader = sqlite_cmd.ExecuteReader();
            }
            else if (gameTitle != null)
            {
                sqlite_cmd.CommandText = "select game_id, title from game where active = 1 and title = @title";
                sqlite_cmd.Parameters.AddWithValue("@title", gameTitle);
                sqlite_datareader = sqlite_cmd.ExecuteReader();
            }
            else
            {
                Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + ": Either the game id or title have to be provided." + Environment.NewLine);
                sqlite_conn.Close();
                return;
            }

            int count = 0;
            while (sqlite_datareader.Read()) // Read() returns true if there is still a result line to read
            {

                object game_id;
                string title;
                int syncState = 0;

                game_id = sqlite_datareader.GetValue(0);
                title = sqlite_datareader.GetString(1);

                Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + ": " + title + " ");

                SQLiteCommand sqlite_cmd2;
                SQLiteDataReader sqlite_datareader2;

                sqlite_cmd2 = sqlite_conn.CreateCommand();
                sqlite_cmd2.CommandText = "select savegame_id, path, machine, recursive, filter from savegame where game_id = @gameid";
                sqlite_cmd2.Parameters.Add("@gameid", (DbType)SqlDbType.Int).Value = game_id;
                sqlite_datareader2 = sqlite_cmd2.ExecuteReader();

                while (sqlite_datareader2.Read())
                {
                    int savegame_id;
                    string path;
                    int machine;
                    int recursive;
                    string filter = null;

                    savegame_id = (int)(long)sqlite_datareader2.GetValue(0);
                    path = sqlite_datareader2.GetString(1);
                    path = ReplaceEnvironmentVariables(path);
                    machine = (int)(long)sqlite_datareader2.GetValue(2);
                    recursive = (int)(long)sqlite_datareader2.GetValue(3);
                    try
                    {
                        filter = sqlite_datareader2.GetString(4);
                    } catch { }                    

                    if (machine == 1)
                    {
                        recursive = 1;
                        filter = null;
                    }

                    var gameFolder = driveHelper.GetFolder(gamesaveRootFolder.Id, game_id.ToString());
                    ICloudFile gamesaveFolder = null;

                    if (gameFolder is not null)
                    {
                        gamesaveFolder = driveHelper.GetFolder(gameFolder.Id, savegame_id.ToString());
                        if (gamesaveFolder is not null)
                        {
                            if (!Directory.Exists(path))
                            {
                                Directory.CreateDirectory(path);
                            }
                        }
                        else if (Directory.Exists(path))
                        {
                            gamesaveFolder = driveHelper.NewFolder(savegame_id.ToString(), gameFolder.Id);
                        }
                    }
                    else if (Directory.Exists(path))
                    {
                        gameFolder = driveHelper.NewFolder(game_id.ToString(), gamesaveRootFolder.Id);
                        gamesaveFolder = driveHelper.NewFolder(savegame_id.ToString(), gameFolder.Id);
                    }

                    if (gamesaveFolder is not null)
                    {

                        int syncResult = SyncPath(path, gamesaveFolder.Id, performBackup, game_id.ToString() + "_" + savegame_id.ToString() + "_" + DateTime.Now.ToString("ddMMyyyy_HHmmss"), recursive==1, filter, syncDirection);
                        if (syncResult == -1)
                        {
                            if (machine == 1)
                            {
                                // update machine subfolders with the most recent when synchronizing from cloudService
                                UpdateMachine(path);
                            }
                            // sync from cloudService
                            syncState = 3;
                        }
                        else if (syncResult == 1)
                        {
                            // sync from local computer
                            syncState = 2;
                        }
                        else if (syncResult == 0)
                        {
                            if (syncState == 0)
                            {
                                // already sync
                                syncState = 1;
                            }
                        }

                    }

                }
                sqlite_datareader2.Close();

                switch (syncState)
                {
                    case 0:
                        {
                            // skipped: not in local and not in cloudService and game active
                            Log("skipped" + Environment.NewLine);
                            break;
                        }
                    case 1:
                        {
                            Log("in sync" + Environment.NewLine);
                            break;
                        }
                    case 2:
                        {
                            Log("synched from local computer" + Environment.NewLine);
                            break;
                        }
                    case 3:
                        {
                            Log("synched from " + cloudService + Environment.NewLine);
                            break;
                        }
                }
                count++;
            }
            sqlite_datareader.Close();
            sqlite_conn.Close();

            if (count == 0)
            {
                Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + $": No games found in the database with the specified criteria. ID={gameId} TITLE={gameTitle}" + Environment.NewLine);
                sqlite_conn.Close();
                return;
            }

            var endTime = DateTime.Now;
            double secs = endTime.Subtract(startTime).TotalSeconds;
            Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + $": *** Synchronization completed in {secs:N2} seconds" + Environment.NewLine);

        }


        public void Close()
        {
            var sessionEndTime = DateTime.Now;
            double secs = sessionEndTime.Subtract(sessionStartTime).TotalSeconds;
            Log(DateTime.Now.ToString("dd/MM/yyyy HH:mm:ss") + $": Total session time {secs:N2} seconds" + Environment.NewLine);
            Log(Environment.NewLine);

            logFile?.Close();

        }


        public void Log(string text)
        {
            Console.Write(text);
            logFile?.Write(text);
        }

        private static void BinaryStreamCopy(Stream sReader, Stream sWriter)
        {

            // Create a read buffer. The buffer length is less than or equal to the data length
            long dataLength = sReader.Length;
            int bufferLength = (int)(dataLength >= 1024L ? 1024L : dataLength);
            byte[] buffer = new byte[bufferLength];
            while (true)
            {
                int read = sReader.Read(buffer, 0, buffer.Length);
                if (read == 0)
                    break;
                sWriter.Write(buffer, 0, read);
            }

        }

        // Fully synchronizes a local and could folder
        // Result: 0 no synch needed, 1 synchronizes from local, -1 synchronizes from cloudService
        private int SyncPath(string folderPath, string folderId, bool backup, string backupName, bool recursive, string filter, string syncDirection)
        {
            int syncResult = 0;

            switch (syncDirection)
            {
                case "tocloud":
                    driveHelper.SyncFromLocal(folderPath, folderId, recursive, filter);
                    syncResult = 1;
                    break;
                case "fromcloud":
                    if (backup)
                    {
                        ZipFile.CreateFromDirectory(folderPath, Path.Combine(backupFolder, backupName) + ".zip", CompressionLevel.Optimal, true);
                    }
                    driveHelper.SyncFromDrive(folderPath, folderId, recursive, filter);
                    syncResult = -1;
                    break;
                default:
                    int totalFilesDrive = 0;
                    var lastModifiedDrive = driveHelper.LastModifiedDate(folderId, ref totalFilesDrive, recursive, filter);
                    int totalFilesLocal = 0;
                    var lastModifiedLocal = driveHelper.LocalLastModifiedDate(folderPath, ref totalFilesLocal, recursive, filter);

                    if (lastModifiedDrive != default)
                    {
                        if (lastModifiedLocal != default)
                        {
                            if (lastModifiedDrive.ToString("yyyyMMdd HHmmss") != lastModifiedLocal.ToString("yyyyMMdd HHmmss") || totalFilesDrive != totalFilesLocal)
                            {
                                if (lastModifiedDrive > lastModifiedLocal)
                                {
                                    if (backup)
                                    {
                                        ZipFile.CreateFromDirectory(folderPath, Path.Combine(backupFolder, backupName) + ".zip", CompressionLevel.Optimal, true);
                                    }
                                    driveHelper.SyncFromDrive(folderPath, folderId, recursive, filter);
                                    syncResult = -1;
                                }
                                else
                                {
                                    driveHelper.SyncFromLocal(folderPath, folderId, recursive, filter);
                                    syncResult = 1;
                                }
                            }
                        }
                        else
                        {
                            if (backup)
                            {
                                ZipFile.CreateFromDirectory(folderPath, Path.Combine(backupFolder, backupName) + ".zip", CompressionLevel.Optimal, true);
                            }
                            driveHelper.SyncFromDrive(folderPath, folderId, recursive, filter);
                            syncResult = -1;
                        }
                    }
                    else if (lastModifiedLocal != default)
                    {
                        driveHelper.SyncFromLocal(folderPath, folderId, recursive, filter);
                        syncResult = 1;
                    }
                    break;
            }

            return syncResult;

        }

        private static string ReplaceEnvironmentVariables(string sPath)
        {

            int pos1 = sPath.IndexOf("%");
            if (pos1 == -1)
            {
                return sPath;
            }
            int pos2 = sPath.IndexOf("%", pos1 + 1);
            if (pos2 == -1)
            {
                return sPath;
            }

            string variable = sPath.Substring(pos1, pos2 - pos1 + 1);
            return sPath.Replace(variable, Environment.GetEnvironmentVariable(variable.Replace("%", "")));

        }

        private void UpdateMachine(string folderPath)
        {

            DateTime lastModified = default;
            string lastModifiedPath = null;
            string[] folderEntries = Directory.GetDirectories(folderPath);
            foreach (var folderEntry in folderEntries)
            {
                int totalFiles = 0;
                var folderLastModified = driveHelper.LocalLastModifiedDate(folderEntry, ref totalFiles, true, null);
                if (lastModified == default || folderLastModified > lastModified)
                {
                    lastModified = folderLastModified;
                    lastModifiedPath = folderEntry;
                }
            }

            foreach (var folderEntry in folderEntries)
            {
                if (folderEntry != lastModifiedPath)
                {
                    Directory.Delete(folderEntry, true);
                    Directory.CreateDirectory(folderEntry);
                    Microsoft.VisualBasic.FileIO.FileSystem.CopyDirectory(lastModifiedPath, folderEntry, true);
                }
            }

        }


    }
}